packages/core/src/odata/v4/payload-value-converter.ts

```diff
@@ -40,117 +43,74 @@ type EdmTypeMapping = {
-export function edmDateTimeToMoment(edmDateTime: string): Moment {
-  const dateTimeOffsetComponents = /.*\((-?\d+)(?:([\+-])(\d{4})\))?/.exec(
-    edmDateTime
-  );
-  if (!dateTimeOffsetComponents) {
-    throw new Error(`Failed to parse edmDateTime: ${edmDateTime} to moment.`);
-  }
-
-  const timestamp = moment(parseInt(dateTimeOffsetComponents[1]));
-
-  if (dateTimeOffsetComponents[2] && dateTimeOffsetComponents[3]) {
-    const offsetMultiplier = dateTimeOffsetComponents[2] === '+' ? 1 : -1;
-    const offsetInMinutes = parseInt(dateTimeOffsetComponents[3]);
-
-    return timestamp.utc().utcOffset(offsetMultiplier * offsetInMinutes);
-  }
-
-  return timestamp;
+  return parsed;
 }
 
-/**
- * @hidden
- * This function can be used for both Edm.DateTime and and Edm.DateTimeOffset.
- */
-export function momentToEdmDateTime(momentInstance: Moment): string {
-  const timestamp = momentInstance.unix() * 1000;
-
-  // For some reason isUtc() returns wrong values here, so we use the internal flag directly
-  if (momentInstance['_isUTC']) {
-    const offset = Math.abs(momentInstance.utcOffset());
-    const operator = momentInstance.utcOffset() >= 0 ? '+' : '-';
-    return `/Date(${timestamp}${operator}${leftpad(offset, 4)})/`;
+function edmDateTimeOffsetToMoment(dateTime: string): Moment {
+  const prefix = 'YYYY-MM-DDTHH:mm';
+  // In moment the Z is either Offset from UTC as +-HH:mm, +-HHmm, or Z
+  const validFormats = [`${prefix}Z`, `${prefix}:ssZ`, `${prefix}:ss.SSSZ`];
+  const parsed = moment(dateTime, validFormats, true);
+  if (!parsed.isValid()) {
+    throw new Error(
+      `Provided date-time value ${dateTime} does not follow the Edm.DateTimeOffset pattern: YYYY-MM-DDTHH:mm(:ss(.SSS))Z`
+    );
   }
-
-  return `/Date(${timestamp})/`;
+  return parsed;
 }
 
-/**
- * @hidden
- */
-export function parseNumber(value: string | number): number {
-  if (typeof value === 'number') {
-    return value;
+function edmDurationToMoment(value: string): Duration {
+  const durationPattern = /([\+,\-]{1,1})?P(\d+D)?T(\d{1,2}H)?(\d{1,2}M)?(\d{1,2}S)?(\d{2,2}\.\d+S)?/;
+  const captured = durationPattern.exec(value);
+  if (!captured || captured[0] !== value) {
+    throw new Error(
+      `Provided duration value ${value} does not follow the Edm.Duration pattern: +/- P0DT0H0M0S`
+    );
   }
+  return moment.duration(value);
+}
 
-  if (value.toLowerCase() === 'inf') {
-    return Number.POSITIVE_INFINITY;
-  }
-  if (value.toLowerCase() === '-inf') {
-    return Number.NEGATIVE_INFINITY;
-  }
-  if (value.toLowerCase() === 'nan') {
-    return Number.NaN;
+function edmTimeOfDay(value: string): Moment {
+  const prefix = 'HH:mm:ss';
+  const validFormats = [`${prefix}`, `${prefix}.SSS`];
+  const parsed = moment.utc(value, validFormats, true);
+  if (!parsed.isValid()) {
+    throw new Error(
+      `Provided time-of-day value ${value} does not follow the Edm.TimeOfDay pattern: HH:mm:ss(.SSS)`
+    );
   }
+  return moment.utc({
+    y: 1970,
```

Comment:
Ouf. Isn't that why we have `Time`? Shouldn't we rather use that?

Reply:
oh, yes good point. I created a extension to time containing fractional seconds.